src/core/model/table/dynamic_position_list_index.h

```diff
@@ -0,0 +1,108 @@
+    }
+};
+
+namespace model {
+class DynamicPositionListIndex {
+public:
+    using Cluster = std::vector<int>;
+
+private:
+    std::unordered_map<std::vector<int>, Cluster> clusters_;
+    unsigned int size_;
+    unsigned int relation_size_;
+    unsigned int original_relation_size_;
+    std::shared_ptr<std::vector<int> const> probing_table_cache_;
+    mutable std::vector<std::vector<int>> cluster_values_cache_;
+
+public:
+    static int intersection_count_;
+    static unsigned long long micros_;
+    static int const kSingletonValueId;
+
+    DynamicPositionListIndex(std::unordered_map<std::vector<int>, Cluster> index, unsigned int size,
+                             unsigned int relation_size, unsigned int original_relation_size);
+
+    static std::unique_ptr<DynamicPositionListIndex> CreateFor(std::vector<std::vector<int>>& data);
+
+    void UpdateWith(std::vector<std::vector<int>>& insert_data,
+                    std::vector<size_t> const& delete_data);
+
+    static std::unordered_map<int, unsigned> CreateFrequencies(
+            Cluster const& cluster, std::vector<int> const& probing_table);
+
+    // если PT закеширована, выдаёт её, иначе предварительно вычисляет её -- тяжёлая операция
+    std::shared_ptr<std::vector<int> const> CalculateAndGetProbingTable() const;
+
+    // выдаёт закешированную PT, либо nullptr, если она не закеширована
+    std::vector<int> const* GetCachedProbingTable() const {
+        return probing_table_cache_.get();
+    };
+
+    // кеширует PT
+    void ForceCacheProbingTable() {
+        probing_table_cache_ = CalculateAndGetProbingTable();
+    };
+
+    std::unordered_map<std::vector<int>, Cluster> const& GetClusters() const noexcept {
+        return clusters_;
+    };
+
+    /* If you use this method and change index in any way, all other methods will become invalid */
+    std::unordered_map<std::vector<int>, Cluster>& GetClusters() noexcept {
+        return clusters_;
+    }
+
+    unsigned int GetNumNonSingletonCluster() const {
+        return clusters_.size();
+    }
+
+    unsigned int GetNumCluster() const {
+        return clusters_.size() + original_relation_size_ - size_;
+    }
+
+    unsigned int GetSize() const {
+        return size_;
+    }
+
+    unsigned int GetRelationSize() const {
+        return relation_size_;
+    }
+
+    std::vector<std::vector<int>> const& GetClusterValuesCache() const {
+        return cluster_values_cache_;
+    }
+
+    bool AllValuesAreUnique() const noexcept {
+        return GetNumNonSingletonCluster() == 0;
+    }
+
+    bool IsConstant() const {
+        return relation_size_ <= 1 || (GetNumNonSingletonCluster() == 1 && size_ == relation_size_);
+    }
+
+    std::unique_ptr<DynamicPositionListIndex> Intersect(DynamicPositionListIndex const* that) const;
+    std::unique_ptr<DynamicPositionListIndex> Probe(DynamicPositionListIndex const* that) const;
+    std::string ToString() const;
+};
+
+using DynPLI = DynamicPositionListIndex;
```

Comment:
Not useful alias. [Relevant](https://google.github.io/styleguide/cppguide.html#Aliases)